src/dump.c

```diff
@@ -1082,25 +1082,45 @@ static void jl_serialize_value_(jl_serializer_state *s, jl_value_t *v, int as_li
         }
 
         char *data = (char*)jl_data_ptr(v);
-        size_t i, np = t->layout->npointers;
-        const char *start = data;
-        for (i = 0; i < np; i++) {
-            uint32_t ptr = jl_ptr_offset(t, i);
-            jl_value_t **fld = &((jl_value_t**)data)[ptr];
-            if ((const char*)fld != start)
-                ios_write(s->s, start, (const char*)fld - start);
-            jl_value_t *e = *fld;
-            JL_GC_PROMISE_ROOTED(e);
-            if (t->mutabl && e && jl_is_cpointer(e) && jl_unbox_voidpointer(e) != (void*)-1 && jl_unbox_voidpointer(e) != NULL)
-                // reset Ptr fields to C_NULL (but keep MAP_FAILED / INVALID_HANDLE)
-                jl_serialize_cnull(s, jl_typeof(e));
-            else
-                jl_serialize_value(s, e);
-            start = (const char*)&fld[1];
-        }
-        data += jl_datatype_size(t);
-        if (data != start)
-            ios_write(s->s, start, data - start);
+        size_t i, j, np = t->layout->npointers;
+        uint32_t nf = t->layout->nfields;
+        char *last = data;
+        for (i = 0, j = 0; i < nf+1; i++) {
+            char *ptr = data + (i < nf ? jl_field_offset(t, i) : jl_datatype_size(t));
+            if (j < np) {
+                char *prevptr = (char*)&((jl_value_t**)data)[jl_ptr_offset(t, j)];
+                while (ptr > prevptr) {
+                    // previous field contained pointers; write them and their interleaved data
+                    if (prevptr > last)
+                        ios_write(s->s, last, prevptr - last);
+                    jl_value_t *e = *(jl_value_t**)prevptr;
+                    JL_GC_PROMISE_ROOTED(e);
+                    if (t->mutabl && e && jl_is_cpointer(e) && jl_unbox_voidpointer(e) != (void*)-1 && jl_unbox_voidpointer(e) != NULL)
```

Comment:
You probably want `jl_field_type(t, i)->mutabl` here (but that's always false)
```suggestion
                    if (0)
```

Reply:
Oh right, we need some kind of check for whether this is a direct field of the object. Maybe checking that `jl_field_type(t, i-1)` is concrete and inlinealloc?

Reply:
Ah, right. We probably want `t->mutable && jl_field_isptr(t, i - 1) && jl_is_cpointer(e) &&` then?